### ISSUE TITLE
Incorrect escaping of curly braces in index entries

### ISSUE THREAD

**Issue by cpitclaudel** [sphinx-doc/sphinx#4979]
### Problem

As explained in https://tex.stackexchange.com/questions/153291/index-unmatched-braces-in-latex, the LaTeX `\index` command requires matched braces: it's not correct to write this: `\index{main \{}`.

Unfortunately, that's the code that Sphinx generates for the following example:

```rst
.. index::
   single: main {
```

```latex
\index{main \{}\ignorespaces 
````

This problem happens in real life, ironically, when trying to index a collection of error messages from a compiler. One of the error messages is something like `Unmatched delimiter: {`, which trips Sphinx.

#### Error logs / results

```
Runaway argument?
{main \{}\ignorespaces 
! Paragraph ended before \@wrindex was complete.
<to be read again> 
                   \par 
l.71 
```

#### Expected results

As mentioned in the StackOverflow post above, the solution would be to escape backslashes more aggressively — using a separate macro definition, for example.

### Environment info
- OS: GNU/Linux
- Python version: 3.5.2
- Sphinx version: 1.7.4

**Comment by tk0miya** [comment]
Fixed by #4987.
Thank you for reporting.

**Comment by cpitclaudel** [comment]
Thanks!
